### DISCUSSION TITLE
Roadmap brainstorming

### DISCUSSION THREAD

**Discussion by fmrico** [PlanSys2/ros2_planning_system#152] [Ideas]
Dear PlanSys2 developers, 

I would like to have a brainstorming about the roadmap for the following months. I will propose some improvements, and it would be good to have feedback from you.

- PDDL & Planning
    - [ ] **Evaluate and  Improve PDDL support** 
    - [ ] **Online replanning**: The planner could be trying to calculate new plans continuously, and if a better plan is calculated, switch to the improved plan at runtime.
    - [ ] **Explore a better default planner**
- Tools
    - [ ] **A new GUI**. We could develop different rqt plugins as a complement for Terminal:
        - Problem viewer
        - Problem editor
        - Execution controller
        - Execution monitor
- Test & CI
    - [ ] **Improve CI tests**. I would need help fine-tuning the GitHub actions
    - [ ] **Improve the coverage to 50%**
- Documentation
    - [ ] **Update docs and tutorials** Maybe some of them could be slightly updated.
    - [ ] **Include more tutorials in Web page**
- Multirobot support
    - [ ] **Prepare an example with multiple robots in simulation**
    - [ ] **Experiments with several real robot**
    - [ ] **implement a decentralized execution mechanism**. Currently, only one robot runs PlanSys2 and requests action executions in other robots.

More ideas? @teyssieuman @jjzapf @xydesa

**Comment by jjzapf** [comment]
One small idea ... Predicates and functions are currently stored using C++ standard vectors. It seems like a standard set might be more appropriate for the predicates and a standard map more appropriate for the functions.

**Comment by roveri-marco** [comment]
It would be good to add some debugging feature to e.g. check during execution why an action fails, e.g. conditions do not hold, timeout, ... (it might be this feature is already there, but we were not able to find it).

**Reply by fmrico** [reply]
Yes. Actually, #156 add more verbose runtime errors, but maybe we should work on this...

**Comment by gregtkogut** [comment]
One idea I've had is making a more generic planner plugin that can handle all (most) planners that adhere to the IPC competition pattern for arguments and output.  This occurred to me while developing the [OPTIC plugin](https://github.com/sscpac/optic_planner/tree/main/plansys2_optic_plan_solver) that has about 95% duplicated code from the popf plugin.  And there are several plansys2 plugins out there that appear to be broken because of relatively trivial changes to the PlanSys2 core in the master/galactic-devel branches.     A single, canonical, IPC planner plugin might be easier to maintain and would make adding additional IPC planners (whether those built in workspaces like popf or ROS2-independent) a bit easier.    Thoughts?
